File: 894.All-Possible-Full-Binary-Trees.py
# https://leetcode.com/problems/all-possible-full-binary-trees/description/
#
# algorithms
# Medium (64.1%)
# Total Accepted:    3k
# Total Submissions: 4.7k
# beats 87.00% of python submissions

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def allPossibleFBT(self, N):
        """
        :type N: int
        :rtype: List[TreeNode]
        """
        if N & 1 == 0:
            return []

        if N == 1:
            return [TreeNode(0)]

        res = []
        for i in xrange(1, N / 2 + 1, 2):
            left_arr = self.allPossibleFBT(i)
            right_arr = self.allPossibleFBT(N - 1 - i)
            for left in left_arr:
                for right in right_arr:
                    root = TreeNode(0)
                    root.left = left
                    root.right = right
                    res.append(root)
                    if i != N - i - 1:
                        root = TreeNode(0)
                        root.left = right
                        root.right = left
                        res.append(root)

        return res

File: 894.All-Possible-Full-Binary-Trees.java
// https://leetcode.com/problems/all-possible-full-binary-trees/
//
// algorithms
// Medium (71.71%)
// Total Accepted:    22,376
// Total Submissions: 31,205
// beats 100.0% of java submissions


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<TreeNode> allPossibleFBT(int N) {
        ArrayList<TreeNode> res = new ArrayList<>();
        if (N == 1) {
            TreeNode root = new TreeNode(0);
            res.add(root);
        }

        for (int i = 1; i <= N / 2; i += 2) {
            List<TreeNode> leftArr = allPossibleFBT(i);
            List<TreeNode> rightArr = allPossibleFBT(N - 1 - i);
            boolean isSym = N - 1 - i != i;
                
            for (TreeNode l : leftArr) {
                for (TreeNode r : rightArr) {
                    TreeNode node = new TreeNode(0);
                    node.left = l;
                    node.right = r;
                    res.add(node);
                    if (isSym) {
                        TreeNode node1 = new TreeNode(0);
                        node1.left = r;
                        node1.right = l;
                        res.add(node1);
                    }
                }
            }
        }
        
        return res;
    }
}